### PR TITLE
[Merged by Bors] - feat (LinearAlgebra/FreeProduct): introduce basic structure, universal property

### PR DESCRIPTION
**Pull request by robertmaxton42**
Adds the free product of algebras, together with its universal property, in preparation for invoking it in a future `Category.AlgebraCat.FilteredColimits.lean`. 

---
While I believe `FreeAlgebra R (⨁ (i : ι), A i)` represents the same structure (something I will prove in a future commit), it's harder to state the relevant alg-homs as `FreeAlgebra R` works on arbitrary types, and at any rate I'd still need to provide the actual "product of algebras" interface.

The file is somewhat slow to compile, especially `lift`; I would appreciate advice as to speeding it up.
 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/FreeProduct/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+
+(Special case of the third isomorphism theorem.)-/
+def algEquiv_quot_algEquiv
+    {R : Type u} [CommSemiring R] {A B : Type v} [Semiring A] [Semiring B]
+    [Algebra R A] [Algebra R B] (f : A ≃ₐ[R] B) (rel : A → A → Prop) :
+    RingQuot rel ≃ₐ[R] RingQuot (rel on f.symm) :=
+  AlgEquiv.ofAlgHom
+    (RingQuot.liftAlgHom R (s := rel)
+      ⟨AlgHom.comp (RingQuot.mkAlgHom R (rel on f.symm)) f,
+      λ x y h_rel ↦ by
```
Comment (Ruben-VandeVelde) [inline Mathlib/LinearAlgebra/FreeProduct/Basic.lean:74]: We avoid λ in mathlib:
```suggestion
      fun x y h_rel ↦ by
```

File: Mathlib/LinearAlgebra/FreeProduct/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+        apply RingQuot.mkAlgHom_rel
+        unfold_let rel; simpa [Function.onFun]⟩)
+    ((RingQuot.liftAlgHom R (s := rel on f.symm)
+      ⟨AlgHom.comp (RingQuot.mkAlgHom R rel) f.symm,
+      λ x y h ↦ by apply RingQuot.mkAlgHom_rel; simpa⟩))
```
Comment (Ruben-VandeVelde) [inline Mathlib/LinearAlgebra/FreeProduct/Basic.lean:79]:
```suggestion
      fun x y h ↦ by apply RingQuot.mkAlgHom_rel; simpa⟩))
```

File: Mathlib/LinearAlgebra/FreeProduct/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+/--If two (semi)rings are equivalent and their quotients by a relation `rel` are defined,
+then their quotients are also equivalent.
+
+(Special case of `algEquiv_quot_algEquiv` when `R = ℕ`, which in turn is a special
+case of the third isomorphism theorem.)-/
+def equiv_quot_equiv {A B : Type v} [Semiring A] [Semiring B] (f : A ≃+* B) (rel : A → A → Prop)  :
+    RingQuot rel ≃+* RingQuot (rel on f.symm) :=
+  let f_alg : A ≃ₐ[ℕ] B :=
+    AlgEquiv.ofRingEquiv (f := f) (λ n ↦ by simp)
```
Comment (Ruben-VandeVelde) [inline Mathlib/LinearAlgebra/FreeProduct/Basic.lean:90]:
```suggestion
    AlgEquiv.ofRingEquiv (f := f) (fun n ↦ by simp)
```

File: Mathlib/LinearAlgebra/FreeProduct/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+* `lift` is the universal property of the free product.
+
+## Main results
+
+* `equivPowerAlgebra` establishes an equivalence between `FreeProduct R A` and
+  `FreeProduct_ofPowers R A`.
+* `FreeProduct` is the coproduct in the category of `R`-algebras.
+
+(TODO: Induction principle for `FreeProduct`; show equivalence of `FreeProduct` with
+`PiTensorProduct` in the commutative case and with `FreeAlgebra R ⨁ (i : ι), A i`.)
```
Comment (jcommelin) [inline Mathlib/LinearAlgebra/FreeProduct/Basic.lean:42]:
```suggestion
## TODO
- Induction principle for `FreeProduct`
- Equivalence of `FreeProduct` with `PiTensorProduct` in the commutative case and with `FreeAlgebra R ⨁ (i : ι), A i`
```

**Comment by jcommelin** [comment]
bors d+

**Comment by robertmaxton42** [comment]
bors merge

**Pull request closed without merging**